// SPDX-FileCopyrightText: 2025 Carnegie Mellon University - Living Edge Lab
//
// SPDX-License-Identifier: GPL-2.0-only

syntax = "proto3";
package steeleagle.protocol.services.compute_service;

import "steeleagle_sdk/protocol/common.proto";

/*
 * Used to configure datasinks for sensor streams.
 *
 * This service is used to configure datasink endpoints for frames and
 * telemetry post-processing. It maintains an internal consumer list of
 * datasinks that the kernel broadcasts frames and telemetry to. RPC
 * methods within this service allow for manipulation of this list.
 */
service Compute {
  /*
   * Add datasinks to consumer list.
   *
   * Takes a list of datasinks and adds them to the current consumer list.
   */
  rpc AddDatasinks (AddDatasinksRequest)
	returns (steeleagle.protocol.common.Response) {}
  /*
   * Set the datasink consumer list.
   *
   * Takes a list of datasinks and replaces the current consumer list with them.
   */
  rpc SetDatasinks (SetDatasinksRequest)
	returns (steeleagle.protocol.common.Response) {}
  /*
   * Remove datasinks from consumer list.
   *
   * Takes a list of datasinks and removes them from the current consumer list.
   */
  rpc RemoveDatasinks (RemoveDatasinksRequest)
	returns (steeleagle.protocol.common.Response) {}
}

// Denotes where a datasink is located.
enum DatasinkLocation {
  REMOTE = 0; // remote location (network hop required)
  LOCAL = 1; // local location (IPC)
}

enum InputSource {
  // default value
  SOURCE_UNSPECIFIED = 0;
  // telemetry from the vehicle driver
  DRIVER_TELEMETRY = 1;
  // telemetry from the mission service
  MISSION_TELEMETRY = 2;
  // imagery from the vehicle
  IMAGERY = 3;
}

// Information about a datasink.
message DatasinkInfo {
  // datasink ID
  string id = 1;
  // datasink location
  DatasinkLocation location = 2;
  // input sources for this datasink
  repeated InputSource sources = 3;
}

message AddDatasinksRequest {
  steeleagle.protocol.common.Request request = 1; // request data
  repeated DatasinkInfo datasinks = 2; // name of target datasinks
}

message SetDatasinksRequest {
  steeleagle.protocol.common.Request request = 1; // request data
  repeated DatasinkInfo datasinks = 2; // name of target datasinks
}

message RemoveDatasinksRequest {
  steeleagle.protocol.common.Request request = 1; // request data
  repeated DatasinkInfo datasinks = 2; // name of target datasinks
}
